design/authenticators/authn_k8s_error_solution.md

```diff
@@ -0,0 +1,54 @@
+**Note**: This design document has been ported from the original document
+[here](https://github.com/cyberark/conjur/issues/524).
+
+# Background
+The need to write this solution design arises from the problem that was raised as part of secret provider phase II implementation.
+
+The problem is getting error upon authenticate call from secret provider to conjur. The same error appears in all K8S authenticator client usages
+
+# Issue description
+The root cause of the issue that was found appears to be that K8S auth client tries to access to cert file that is written from Conjur K8S Authn by K8S API before this cert file appears / is ready
+
+The reason for this behaviour is that the flow returns to K8S Authn Client without waiting for cert file to be fully written, and K8S Auth client proceeds with flow to read a file - and then we get an error.
+
+# Solution
+The solution should be that K8S Authn Client should wait until and event that says that cert file is ready for use will arrive - and only then it should proceed to read it
+
+In case that the expected event does not arrive after timeout - error must be issued and the flow will proceed without reading cert file
+
+# Design
+
+![Sequence Diagram](images/authenticators_bug_sequence_diagram.jpeg)
+
+The addition to the current design is WaitUntilWrittenOrTimeout call. It should be executed per new flag sent into the InjectClientCert command (in order to stay backward compatible)
```

Comment:
What do you mean "executed per flag"
The design is really important to go into detail so please add a few more sentences explaining the flow

Reply:
@oburstein-hub why is adding this wait is breaking the behavior that was before? in what sense?

Reply:
After a fix system will always wait for file to appear. I wanted to preserve a behaviour when it does not wait. I might be important for some other existing scenarios. If you dont see such - we can eliminate requirement of flag